zebrad/src/commands/start/sync.rs

```diff
@@ -77,61 +92,75 @@ where
         // due to protocol limitations
         self.request_genesis().await?;
 
-        loop {
-            self.obtain_tips().await?;
+        'sync: loop {
+            // Wipe state from prevous iterations.
+            self.prospective_tips = HashSet::new();
+            self.pending_blocks = Box::pin(FuturesUnordered::new());
+
+            tracing::info!("starting sync, obtaining new tips");
+            if self.obtain_tips().await.is_err() {
+                tracing::warn!("failed to obtain tips, waiting to restart sync");
+                delay_for(SYNC_RESTART_TIMEOUT).await;
+                continue 'sync;
+            };
             self.update_metrics();
 
-            // ObtainTips Step 6
-            //
-            // If there are any prospective tips, call ExtendTips.
-            // Continue this step until there are no more prospective tips.
             while !self.prospective_tips.is_empty() {
-                tracing::debug!("extending prospective tips");
-
-                self.extend_tips().await?;
-                self.update_metrics();
+                // Check whether any block tasks are currently ready:
+                while let Some(Some(rsp)) = self.pending_blocks.next().now_or_never() {
```

Comment:
now_or_never??? what is this and where is it defined? (curiousity based question, not worried about it)

Reply:
it's in futures::future::FutureExt

Reply:
sick